### PR TITLE
Fix invalid handling of UTF-8 strings

### PR DESCRIPTION
**Pull request by MrAnno**
This issue can cause invalid UTF-8 characters in `format-cef-extension`, `format-welf` and `format-json`.

`g_utf8_get_char_validated()` converts a sequence of bytes encoded as UTF-8 to a Unicode character (`gunichar` --> UTF-32), so checking for an UTF-8 leading byte (`< 0xc0`) is not an option here.
`0x80` can be used instead.

This is a performance optimization, where we can append the given unichar with `g_string_append_c()` if it fits in 7 bits.
In all other cases, conversion is required (`g_string_append_unichar()`).

Partial fix for #999

### CODE REVIEW

**Comment by MrAnno** [comment]
@bkil-syslogng Can you have a look at this PR, please?

`_append_unichar()` was introduced in [Format CEF extension](https://github.com/balabit/syslog-ng/pull/842/commits), I'm a little bit confused by this `wc < 0xc0` condition.

**Comment by bazsi** [comment]
Anything below 0xC0 can be encoded in a single byte in utf8. This is a performance optimization, that seriously improved performance, and I don't get why it would break things this badly.

**Comment by bkil-syslogng** [comment]
I think the only fix needed is to patch the function to check `(wc < 0x7f)` instead. This one passes the above tests for me.

**Comment by bkil-syslogng** [comment]
Otherwise good catch.

**Comment by MrAnno** [comment]
@bazsi Thank you. I think it's because `wc` is not an UTF-8 encoded character anymore (`g_utf8_get_char_validated()` converts it to a Unicode character).

Is it acceptable if I change 0xc0 to 0x7f as @bkil-syslogng suggested? This should preserve your optimization for ASCII characters.

**Comment by bkil-syslogng** [comment]
Just for the record, we've now discussed with @bazsi in person that such a workaround would be okay if it is really needed.

Here's the source I used to propose `(wc < 0x7f)`:

> Sixty-five code points (U+0000–U+001F and U+007F–U+009F) are reserved as control codes, and correspond to the C0 and C1 control codes defined in ISO/IEC 6429. Of these U+0009 (Tab), U+000A (Line Feed), and U+000D (Carriage Return) are widely used in Unicode-encoded texts.
https://en.wikipedia.org/wiki/Unicode#Character_General_Category

We do escape control characters < 0x20, not sure what results in appending 0x7f directly.

File: lib/utf8utils.c [hunk 1]
```diff
@@ -36,15 +36,6 @@ _is_character_unsafe(gunichar uchar, const gchar *unsafe_chars)
   return _strchr_optimized_for_single_char_haystack(unsafe_chars, (gchar) uchar) != NULL;
```
Comment (bkil-syslogng) [inline lib/utf8utils.c]: Another optimization I see is that we could probably check `(uchar >= 0x7f)` (or `0x80`) in `_is_character_unsafe()`.

**Comment by MrAnno** [comment]
I updated the PR to preserve the mentioned performance optimization. I also fixed the same bug in `format-cef-extension.c`.

**Comment by bkil-syslogng** [comment]
Sorry for the mistake, if I'm reading this correctly, thankfully `0x7F` should be left intact as well, so the check should be `(wc < 0x80)`. The above text was not clear, so we should improve it on Wikipedia.
https://en.wikipedia.org/wiki/C0_and_C1_control_codes#C0_.28ASCII_and_derivatives.29

**Comment by MrAnno** [comment]
@bkil-syslogng U+007F is a control character, why should we append it as is?

**Comment by bkil-syslogng** [comment]
:+1:

**Comment by bazsi** [comment]
This looks good to me and I fully trust @bkil-syslogng about the need to escape 0x7f or not. So if he says it is ok, I agree :)

so  :+1:

**Comment by bkil-syslogng** [comment]
@bazsi **TL;DR**: the reasoning only confirmed that `< 0x80` is the logically sound choice. This PR should fix the said issue of Unicode bytecode breakage in its current form, so it is ready for merging.

Strictly speaking, as `0x7f` is considered a control character, we would probably need to escape it just like `0x00`-`0x1f`. However, this does not affect the said `if`-condition - it's an improvement to be made in the outer loop. Though, because it's an extension to the hot path, we need to be cautious. It would also be wise to research how our users would want to see it. Hence I propose that this possible further enhancement should go into a separate PR.

**Comment by bkil-syslogng** [comment]
@dnsjts Pointed out that the manual inlining could also have been validated from the source (`< 0x80`)
https://git.gnome.org/browse/glib/tree/glib/gstring.c?id=39a22880b6d8544fec51ba3176214f38a8a572b8#n820

**Comment by lbudai** [comment]
Hmm... we have a copy pasted function.
Wouldn't be better to have only one function? Maybe you talked about it and you have your reasons.

@bazsi: how much performance gain can we expect from this? (I'm just curious.)

**Comment by bazsi** [comment]
I can't remember how we ended up with two copies of that function, so that
should probably be eliminated. Maybe @bkil-syslogng remembers.

Doing it with simple append_c instead of append_unichar is a lot faster if
most of the characters are ASCII and they are.

The first is inlined, the 2nd is a function call, in a loop that gets
executed for all characters all affected messages. If we use 20k as message
rate that loop might be executed 2million times per second assuming 100
characters per message.

We do have a bottleneck in format-json as Evan explained in one of his
emails.

**Comment by bkil-syslogng** [comment]
Yes, I remember. I recommended extracting the functionality to a common
place, but I was downvoted.

On Tue, Jun 21, 2016 at 10:50 PM, Balazs Scheidler <notifications@github.com
> wrote:

> I can't remember how we ended up with two copies of that function, so that
> should probably be eliminated. Maybe @bkil-syslogng remembers.
>
> Doing it with simple append_c instead of append_unichar is a lot faster if
> most of the characters are ASCII and they are.
>
> The first is inlined, the 2nd is a function call, in a loop that gets
> executed for all characters all affected messages. If we use 20k as message
> rate that loop might be executed 2million times per second assuming 100
> characters per message.
>
> We do have a bottleneck in format-json as Evan explained in one of his
> emails.

**Comment by lbudai** [comment]
Finally I was able to fix OBS (libdbi and some other deps, libdbi packaging issues, etc...) so Travis should be fine now.

**Comment by lbudai** [comment]
Yeah, the namespace pollution is not the best.

**Pull request merged**